### ISSUE TITLE
Documentation - Movement and state machine

### ISSUE THREAD

**Issue by Ruk33** [Ruk33/l2auth#2]
These are some notes I was thinking about and don't want to lose.

All actions must go through a state machine, otherwise, we may end up with an invalid state or having to make multiple checks before performing an action (ie, attack).

Having said that, we may also want to implement a work queue, but not just for client requests and timer ticks, but for every possible action performed from the server-side. For instance, say we want an NPC to start walking. Instead of immediately notify all the characters close to that NPC, we could add that action (think of it like the server is building and sending a request to itself as the clients do) into a work queue and later process it.

This should, in theory, be easier to reason about, and also, we should be able to split servers up. Why would we want to split servers up? Assume we have 100 players and 100 NPCs. For the server, you will have 200 characters (notice no distinction) all being the same and being processed. It may be good to have a server dedicated just for players so all resources are dedicated to that. I think this is similar to what the original l2off servers do.

In summary:

- All actions that modify state must go through a state machine to prevent invalid states
- All actions are queued up and then processed one by one.

client request -> goes to queue
timer tick -> goes to queue
server-side action (ie, NPC starts to walk) -> goes to queue

```
while (1) {
   process from queue
}
```

Server requests will be the same requests the clients send. So we will have to implement a some_request_to_packet where it converts a request back to raw bytes.

**Comment by Ruk33** [comment]
This is being implemented in https://github.com/Ruk33/l2auth/pull/4

**Comment by Ruk33** [comment]
Implemented in new changes.
